src/util/inotify.c

```diff
@@ -381,44 +381,140 @@ static int watch_ctx_destructor(void *memptr)
     return 0;
 }
 
+static errno_t resolve_filename(TALLOC_CTX *mem_ctx,
+                                struct snotify_ctx *snctx,
+                                const char *filename,
+                                char **out_resolved)
+{
+    TALLOC_CTX *tmp_ctx;
+    char *resolved;
+    errno_t ret;
+
+    tmp_ctx = talloc_new(mem_ctx);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    resolved = talloc_size(tmp_ctx, PATH_MAX + 1);
```

Comment:
I think instead of dynamically allocating a buffer of fixed size using just `char resolved[PATH_MAX + 1];` would be fine here as well, same with `fcopy`. Additionally using `strncpy()` as in the old `copy_filenames()` would help to avoid all the NULL-checks after copying making the code shorter and more readable.

In general I'm all in favor of dynamical memory allocation with talloc, but in the case of file-system paths where all involved glibc calls have a limit at PATH_MAX I find fixed arrays more suitable. But maybe I'm missing something here, please let me know.

Reply:
I generally prefer talloc but I don't have problem reverting copy_filenames back to fixed arrays and change resolve_filename to use them too. The diff is smaller this way.

I have also added a comment to note that GNU extensions to resolvepath are used, it took me a while to figure it out reading the error code path again, otherwise the content of resolved array would be undefined.